miniapps/shifted/extrapolator.cpp

```diff
@@ -0,0 +1,561 @@
+         {
+            dx.SetSubVector(dofs, 0.0);
+            continue;
+         }
+      }
+      return;
+   }
+
+   Vector rhs(x.Size());
+   K_mat->Mult(x, rhs);
+   rhs += b;
+
+   DenseMatrix M_loc(nd);
+   DenseMatrixInverse M_loc_inv(&M_loc);
+   Vector rhs_loc(nd), dx_loc(nd);
+   for (int k = 0; k < NE; k++)
+   {
+      pfes.GetElementDofs(k, dofs);
+
+      if (active_zones[k] == false)
+      {
+         dx.SetSubVector(dofs, 0.0);
+         continue;
+      }
+
+      rhs.GetSubVector(dofs, rhs_loc);
+      M.SpMat().GetSubMatrix(dofs, dofs, M_loc);
+      M_loc_inv.Factor();
+      M_loc_inv.Mult(rhs_loc, dx_loc);
+      dx.SetSubVector(dofs, dx_loc);
+   }
+}
+
+void AdvectionOper::ComputeElementsMinMax(const ParGridFunction &gf,
+                                          Vector &el_min, Vector &el_max) const
+{
+   ParFiniteElementSpace &pfes = *gf.ParFESpace();
+   const int NE = pfes.GetNE(), ndof = pfes.GetFE(0)->GetDof();
+   for (int k = 0; k < NE; k++)
+   {
+      el_min(k) = numeric_limits<double>::infinity();
+      el_max(k) = -numeric_limits<double>::infinity();
+
+      for (int i = 0; i < ndof; i++)
+      {
+         el_min(k) = min(el_min(k), gf(k*ndof + i));
+         el_max(k) = max(el_max(k), gf(k*ndof + i));
+      }
+   }
+}
+
+void AdvectionOper::ComputeBounds(const ParFiniteElementSpace &pfes,
+                                  const Vector &el_min, const Vector &el_max,
+                                  Vector &dof_min, Vector &dof_max) const
+{
+   ParMesh *pmesh = pfes.GetParMesh();
+   L2_FECollection fec_bounds(0, pmesh->Dimension());
+   ParFiniteElementSpace pfes_bounds(pmesh, &fec_bounds);
+   ParGridFunction el_min_gf(&pfes_bounds), el_max_gf(&pfes_bounds);
+   const int NE = pmesh->GetNE(), ndofs = dof_min.Size() / NE;
+
+   el_min_gf = el_min;
+   el_max_gf = el_max;
+
+   el_min_gf.ExchangeFaceNbrData(); el_max_gf.ExchangeFaceNbrData();
+   const Vector &min_nbr = el_min_gf.FaceNbrData();
+   const Vector &max_nbr = el_max_gf.FaceNbrData();
+   const Table &el_to_el = pmesh->ElementToElementTable();
+   Array<int> face_nbr_el;
+   for (int k = 0; k < NE; k++)
+   {
+      double k_min = el_min_gf(k), k_max = el_max_gf(k);
+
+      el_to_el.GetRow(k, face_nbr_el);
+      for (int n = 0; n < face_nbr_el.Size(); n++)
+      {
+         if (face_nbr_el[n] < NE)
+         {
+            // Local neighbor.
+            k_min = std::min(k_min, el_min_gf(face_nbr_el[n]));
+            k_max = std::max(k_max, el_max_gf(face_nbr_el[n]));
+         }
+         else
+         {
+            // MPI face neighbor.
+            k_min = std::min(k_min, min_nbr(face_nbr_el[n] - NE));
+            k_max = std::max(k_max, max_nbr(face_nbr_el[n] - NE));
+         }
+      }
+
+      for (int j = 0; j < ndofs; j++)
+      {
+         dof_min(k*ndofs + j) = k_min;
+         dof_max(k*ndofs + j) = k_max;
+      }
+   }
+}
+
+void Extrapolator::Extrapolate(Coefficient &level_set,
```

Comment:
Can we rewrite the implementation of the Extrapolate method by means of a recursive loop? I think this could a nice way to generalize the method to extrapolation order.

Reply:
We could, but I think it would be too complicated. The same can be done with a loop that calls `TimeLoop(...)`. Still the iterations are not completely the same, especially with `Bochkov`, and in practice I don't expect anyone to need extrapolation above 1st / 2nd order.